jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java

```diff
@@ -92,17 +80,91 @@ public void testExtraDirectoryLayerConfiguration() throws URISyntaxException, IO
             (ignored1, ignored2) -> Instant.EPOCH);
-    assertSourcePathsUnordered(
-        Arrays.asList(
+    assertThat(layerConfiguration.getEntries())
+        .comparingElementsUsing(SOURCE_FILE_OF)
+        .containsExactly(
             extraFilesDirectory.resolve("a"),
             extraFilesDirectory.resolve("a/b"),
             extraFilesDirectory.resolve("a/b/bar"),
             extraFilesDirectory.resolve("c"),
             extraFilesDirectory.resolve("c/cat"),
-            extraFilesDirectory.resolve("foo")),
-        layerConfiguration.getEntries());
+            extraFilesDirectory.resolve("foo"));
+  }
+
+  @Test
+  public void testExtraDirectoryLayerConfiguration_includes()
+      throws URISyntaxException, IOException {
+    Path extraFilesDirectory = Paths.get(Resources.getResource("core/layer").toURI());
+    FileEntriesLayer layerConfiguration =
+        JavaContainerBuilderHelper.extraDirectoryLayerConfiguration(
+            extraFilesDirectory,
+            AbsoluteUnixPath.get("/"),
+            Arrays.asList("**/bar", "**/*a*"),
+            Collections.emptyList(),
+            Collections.emptyMap(),
+            (ignored1, ignored2) -> Instant.EPOCH);
+    assertThat(layerConfiguration.getEntries())
+        .comparingElementsUsing(SOURCE_FILE_OF)
+        .containsExactly(
+            extraFilesDirectory.resolve("a"),
+            extraFilesDirectory.resolve("a/b/bar"),
+            extraFilesDirectory.resolve("c/cat"));
+  }
+
+  @Test
+  public void testExtraDirectoryLayerConfiguration_excludes()
+      throws URISyntaxException, IOException {
+    Path extraFilesDirectory = Paths.get(Resources.getResource("core/layer").toURI());
+    FileEntriesLayer layerConfiguration =
+        JavaContainerBuilderHelper.extraDirectoryLayerConfiguration(
+            extraFilesDirectory,
+            AbsoluteUnixPath.get("/"),
+            Collections.emptyList(),
+            Arrays.asList("**/bar", "**/*a*"),
+            Collections.emptyMap(),
+            (ignored1, ignored2) -> Instant.EPOCH);
+    assertThat(layerConfiguration.getEntries())
+        .comparingElementsUsing(SOURCE_FILE_OF)
+        .containsExactly(
+            extraFilesDirectory.resolve("a/b"),
+            extraFilesDirectory.resolve("c"),
+            extraFilesDirectory.resolve("foo"));
+  }
+
+  @Test
+  public void testExtraDirectoryLayerConfiguration_includesAndExcludesEverything()
+      throws URISyntaxException, IOException {
+    Path extraFilesDirectory = Paths.get(Resources.getResource("core/layer").toURI());
+    FileEntriesLayer layerConfiguration =
+        JavaContainerBuilderHelper.extraDirectoryLayerConfiguration(
+            extraFilesDirectory,
+            AbsoluteUnixPath.get("/"),
+            Arrays.asList("**/*"),
+            Arrays.asList("**/*"),
+            Collections.emptyMap(),
+            (ignored1, ignored2) -> Instant.EPOCH);
+    assertThat(layerConfiguration.getEntries()).isEmpty();
```

Comment:
ah so excludes takes precedence?

Reply:
Yeah, if something is excluded, then that's absolute and final. I just need to double-check the behavior with the maven-resource-plugin.

Reply:
Just verified that the behavior is the same as the Maven resources plugin.